cli/manage.go

```diff
@@ -132,81 +140,71 @@ func getDiscoveryOpt(c *cli.Context) map[string]string {
-	kvDiscovery, ok := discovery.(*kvdiscovery.Discovery)
-	if !ok {
-		log.Fatal("Leader election is only supported with consul, etcd and zookeeper discovery.")
-	}
-	client := kvDiscovery.Store()
-	p := path.Join(kvDiscovery.Prefix(), leaderElectionPath)
-
-	candidate := leadership.NewCandidate(client, p, addr, leaderTTL)
-	follower := leadership.NewFollower(client, p)
-	return candidate, follower
-}
-
-func setupReplication(c *cli.Context, cluster cluster.Cluster, server *api.Server, candidate *leadership.Candidate, follower *leadership.Follower, addr string, tlsConfig *tls.Config) {
-	primary := api.NewPrimary(cluster, tlsConfig, &statusHandler{cluster, candidate, follower}, c.GlobalBool("debug"), c.Bool("cors"))
+func setupReplication(c *cli.Context, cluster cluster.Cluster, client store.Store, keyPath string, server *api.Server, candidate *leadership.Candidate, addr string, tlsConfig *tls.Config) {
+	handler := &statusHandler{cluster, candidate, nil, addr}
+	primary := api.NewPrimary(cluster, tlsConfig, handler, c.GlobalBool("debug"), c.Bool("cors"))
 	replica := api.NewReplica(primary, tlsConfig)
 
 	go func() {
 		for {
-			run(cluster, candidate, server, primary, replica)
-			time.Sleep(defaultRecoverTime)
-		}
-	}()
-
-	go func() {
-		for {
-			follow(follower, replica, addr)
-			time.Sleep(defaultRecoverTime)
+			err := run(cluster, client, keyPath, handler, server, primary, replica, addr)
+			if err != nil {
+				time.Sleep(defaultRecoverTime)
+			}
 		}
 	}()
 
 	server.SetHandler(primary)
 }
 
-func run(cl cluster.Cluster, candidate *leadership.Candidate, server *api.Server, primary *mux.Router, replica *api.Replica) {
-	electedCh, errCh := candidate.RunForElection()
+func run(cl cluster.Cluster, client store.Store, keyPath string, handler *statusHandler, server *api.Server, primary *mux.Router, replica *api.Replica, addr string) error {
+	candidate := handler.candidate
+	electedCh, candidateErrCh := candidate.RunForElection()
+
+	follower := leadership.NewFollower(client, keyPath)
+	defer follower.Stop()
+	handler.follower = follower
+	leaderCh, followerErrCh := follower.FollowElection()
+
 	var watchdog *cluster.Watchdog
+	wasLeader := false
 	for {
 		select {
 		case isElected := <-electedCh:
 			if isElected {
 				log.Info("Leader Election: Cluster leadership acquired")
+				wasLeader = true
 				watchdog = cluster.NewWatchdog(cl)
 				server.SetHandler(primary)
 			} else {
 				log.Info("Leader Election: Cluster leadership lost")
 				cl.UnregisterEventHandler(watchdog)
 				server.SetHandler(replica)
+				if wasLeader {
+					return nil
+				}
 			}
 
-		case err := <-errCh:
-			log.Error(err)
-			return
-		}
-	}
-}
-
-func follow(follower *leadership.Follower, replica *api.Replica, addr string) {
-	leaderCh, errCh := follower.FollowElection()
-	for {
-		select {
 		case leader := <-leaderCh:
 			if leader == "" {
 				continue
 			}
 			if leader == addr {
 				replica.SetPrimary("")
+				server.SetHandler(primary)
```

Comment:
`electedCh` gets triggered when the manager successfully writes out to the KV store or when it loses the connection to the KV store (it returns true in the former case and false in the latter). `leaderCh` gets triggered any time the KV store's value changes. Thus, when a node becomes leader, both channels get triggered.